src/Bicep.LangServer/Handlers/InsertResourceHandler.cs

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Deployments.Core.Comparers;
+using Azure.Deployments.Core.Definitions.Identifiers;
+using Bicep.Core.CodeAction;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Parsing;
+using Bicep.Core.Resources;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem.Az;
+using Bicep.LanguageServer.CompilationManager;
+using Bicep.LanguageServer.Extensions;
+using MediatR;
+using OmniSharp.Extensions.JsonRpc;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Workspace;
+using Bicep.Core.Extensions;
+using Bicep.Core.PrettyPrint;
+using Bicep.Core.PrettyPrint.Options;
+using Bicep.Core.Workspaces;
+using Bicep.LanguageServer.Utils;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using Bicep.Core.Navigation;
+using Bicep.Core.Rewriters;
+using System.Text.RegularExpressions;
+using OmniSharp.Extensions.LanguageServer.Protocol.Window;
+using Bicep.LanguageServer.Providers;
+using Bicep.LanguageServer.Telemetry;
+
+namespace Bicep.LanguageServer.Handlers
+{
+    [Method("textDocument/insertResource", Direction.ClientToServer)]
+    public record InsertResourceParams : TextDocumentPositionParams, IRequest
+    {
+        public string? ResourceId { get; init; }
+    }
+
+    public class InsertResourceHandler : IJsonRpcNotificationHandler<InsertResourceParams>
+    {
+        private readonly ILanguageServerFacade server;
+        private readonly ICompilationManager compilationManager;
+        private readonly IAzResourceProvider azResourceProvider;
+        private readonly IAzResourceTypeLoader azResourceTypeLoader;
+        private readonly ITelemetryProvider telemetryProvider;
+
+        public InsertResourceHandler(ILanguageServerFacade server, ICompilationManager compilationManager, IAzResourceProvider azResourceProvider, IAzResourceTypeLoader azResourceTypeLoader, ITelemetryProvider telemetryProvider)
+        {
+            this.server = server;
+            this.compilationManager = compilationManager;
+            this.azResourceProvider = azResourceProvider;
+            this.azResourceTypeLoader = azResourceTypeLoader;
+            this.telemetryProvider = telemetryProvider;
+        }
+
+        public async Task<Unit> Handle(InsertResourceParams request, CancellationToken cancellationToken)
+        {
+            var context = compilationManager.GetCompilation(request.TextDocument.Uri);
+            if (context is null)
+            {
+                return Unit.Value;
+            }
+
+            if (TryParseResourceId(request.ResourceId) is not {} resourceId)
```

Comment:
>TryParseResourceId [](http://example.com/codeflow?start=16&length=18)

What happens if the type of scope in the resource ID doesn't match `targetScope` in the file?

Reply:
The resource is inserted regardless, and the user would be required to fix the error. I thought about this at the time, and it felt like it'd be more annoying to return an error and force the user to fix the targetScope first.

Another possibility I considered is to generate a readonly editor pane rather than inserting in-document - forcing the user to copy+paste the resource into their file. Thoughts?

Reply:
Since there's always Undo in the editor, it should be fine. It's probably a smoother flowing experience if we keep as is. Popping up another tab could be distracting/interrupting.